frame/democracy/src/lib.rs

```diff
@@ -227,39 +227,45 @@ pub trait Trait: system::Trait + Sized {
 
 	/// Period in blocks where an external proposal may not be re-submitted after being vetoed.
 	type CooloffPeriod: Get<Self::BlockNumber>;
+
+	/// The amount of balance that must be deposited per byte of preimage stored.
+	type PreimageByteDeposit: Get<BalanceOf<Self>>;
 }
 
 /// Info regarding an ongoing referendum.
 #[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
-pub struct ReferendumInfo<BlockNumber: Parameter, Proposal: Parameter> {
+pub struct ReferendumInfo<BlockNumber: Parameter, Hash: Parameter> {
 	/// When voting on this referendum will end.
 	end: BlockNumber,
 	/// The proposal being voted on.
-	proposal: Proposal,
+	proposal_hash: Hash,
 	/// The thresholding mechanism to determine whether it passed.
 	threshold: VoteThreshold,
 	/// The delay (in blocks) to wait after a successful referendum before deploying.
 	delay: BlockNumber,
 }
 
-impl<BlockNumber: Parameter, Proposal: Parameter> ReferendumInfo<BlockNumber, Proposal> {
+impl<BlockNumber: Parameter, Hash: Parameter> ReferendumInfo<BlockNumber, Hash> {
 	/// Create a new instance.
 	pub fn new(
 		end: BlockNumber,
-		proposal: Proposal,
+		proposal_hash: Hash,
 		threshold: VoteThreshold,
 		delay: BlockNumber
 	) -> Self {
-		ReferendumInfo { end, proposal, threshold, delay }
+		ReferendumInfo { end, proposal_hash, threshold, delay }
 	}
 }
 
 decl_storage! {
 	trait Store for Module<T: Trait> as Democracy {
 		/// The number of (public) proposals that have been made so far.
 		pub PublicPropCount get(fn public_prop_count) build(|_| 0 as PropIndex) : PropIndex;
-		/// The public proposals. Unsorted.
-		pub PublicProps get(fn public_props): Vec<(PropIndex, T::Proposal, T::AccountId)>;
+		/// The public proposals. Unsorted. The second item is the proposal's hash.
+		pub PublicProps get(fn public_props): Vec<(PropIndex, T::Hash, T::AccountId)>;
+		/// Map of hashes to the proposal preimage, along with who registered it and their deposit.
+		/// The block number is the block at which it was deposited.
+		pub Preimages: map T::Hash => Option<(Vec<u8>, T::AccountId, BalanceOf<T>, T::BlockNumber)>;
```

Comment:
A lot of items in this tuple. Why not use a struct?

Also why use a `Vec<u8>` rather than a `T::Proposal`?

Reply:
using a `Proposal` complicates the economics of `note_preimage`.

basically, proposals can be really big and laborious to `encode` and `decode`; we can work around this by keeping them an opaque blob, at least until we have collected funds and/or know we want to use it. if we don't, then anyone can craft an annoying complex (to encode/decode) proposal and push it in here from an account that doesn't have enough funds to reserve. this would open up a griefing attack, forcing us to attach a large and/or complex weight/cost to calling this function.

keeping it opaque avoids this.